### PR TITLE
New Workflow Management Screen UI using JavaScript

### PR DESCRIPTION
**Pull request by anuprulez**
# Introduction:
The Pull Request presents a new user interface for workflow's home screen built using JavaScript and thereby retires its mako file (list.mako). It displays list of user's workflows in the center panel of the Galaxy as opposed to previous display in full page. 

# Visuals:

### List of workflows
workflow_1

### Search
workflow_2

# Details:
The workflow's table merges the shared and user's own workflows and can be identified by text in the 'owner' column ('You' for user's own workflows and an email id for the shared workflows). Moreover, as an enhancement, it includes a client side search which filters the list of workflows using query typed in the search textbox (query needs at least 3 characters to search).

# Routing:
It uses client side routing in the same way as the user preferences and custom builds screens.

# Credits:
Thanks a lot to @bgruening and @guerler for suggestions and reviews!

Your suggestions are welcome!

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 1]
```diff
@@ -65,6 +65,8 @@ def index(self, trans, **kwd):
             encoded_id = trans.security.encode_id(wf.id)
             item['url'] = url_for('workflow', id=encoded_id)
             item['owner'] = wf.user.username
+            item['latest_workflow_steps'] = len( wf.latest_workflow.steps )
+            item['published'] = wf.published
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: This line is not necessary, `'published'` field is already filled by the `to_dict` method.

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 2]
```diff
@@ -74,6 +76,11 @@ def index(self, trans, **kwd):
             encoded_id = trans.security.encode_id(wf_sa.stored_workflow.id)
             item['url'] = url_for( 'workflow', id=encoded_id )
             item['owner'] = wf_sa.stored_workflow.user.username
+            item['latest_workflow_steps'] = len( wf_sa.stored_workflow.latest_workflow.steps )
+            item['user_email'] = wf_sa.stored_workflow.user.email
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: I think this line should be removed, it's not necessary to expose the user email through the API, the owner should be enough for what you are doing.

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 3]
```diff
@@ -74,6 +76,11 @@ def index(self, trans, **kwd):
...
+            item['user_name'] = wf_sa.stored_workflow.user.username
+            item['slug'] = wf_sa.stored_workflow.slug
+            item['published'] = wf_sa.stored_workflow.published
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: As above, this is not necessary, `'published'` field is already filled by the `to_dict` method.

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        build_workflows: function( response ) {
+            var workflows = [],
+                username = Galaxy.user.attributes.username;
+            _.each( response, function( wf ) {
+                var wf_obj = {};
+                wf_obj.id = wf.id;
+                wf_obj.text = wf.name;
+                wf_obj.workflow_steps = wf.latest_workflow_steps;
+                wf_obj.published = ( wf.published ? "Yes" : "No" );
+                wf_obj.email = wf.user_email ? wf.user_email : "You";
```
Comment (nsoranzo) [inline client/galaxy/scripts/mvc/workflow/workflow.js]: Remove

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+                wf_obj.slug = wf.slug ? wf.slug : "";
+                wf_obj.username = wf.user_name ? wf.user_name : "";
```
Comment (nsoranzo) [inline client/galaxy/scripts/mvc/workflow/workflow.js]:
```javascript
wf_obj.username = wf.owner === username ? "You" : wf.owner;
```

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 6]
```diff
@@ -74,6 +76,11 @@ def index(self, trans, **kwd):
             encoded_id = trans.security.encode_id(wf_sa.stored_workflow.id)
             item['url'] = url_for( 'workflow', id=encoded_id )
             item['owner'] = wf_sa.stored_workflow.user.username
+            item['latest_workflow_steps'] = len( wf_sa.stored_workflow.latest_workflow.steps )
+            item['user_email'] = wf_sa.stored_workflow.user.email
+            item['user_name'] = wf_sa.stored_workflow.user.username
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: That's the same as the owner, remove.

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+                trHtml = trHtml + '<tr>' +
+                             '<td class="wf-td wf-dpd">' +
+                                 '<div class="dropdown">' +
+                                     '<button class="menubutton" type="button" data-toggle="dropdown">' +
+                                         _.escape( wf.text ) + '<span class="caret"></span>' +
+                                     '</button>' +
+                                     self._templateActions( wf ) +
+                                 '</div>' +
+                              '</td>' +
+                              '<td>' + _.escape( wf.email ) +'</td>' +
```
Comment (nsoranzo) [inline client/galaxy/scripts/mvc/workflow/workflow.js]: s/email/username/

**Review by nsoranzo — COMMENTED** [review]
Another mako gone, thanks a lot @anuprulez!

**Comment by bgruening** [comment]
@nsoranzo @erasche if you have time can you test this. The UI changes a little bit and we would like to have some more feedback on this side as well! And thanks for the review Nicola!

**Comment by hexylena** [comment]
yep, will do! This looks amazing

**Comment by mvdbeek** [comment]
I just tested this out, I think this is a very nice improvement. The fact that this is rendered in the center panel is great, everything works, I'm impressed.

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 8]
```diff
@@ -65,6 +65,7 @@ def index(self, trans, **kwd):
             encoded_id = trans.security.encode_id(wf.id)
             item['url'] = url_for('workflow', id=encoded_id)
             item['owner'] = wf.user.username
+            item['latest_workflow_steps'] = len( wf.latest_workflow.steps )
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: `latest_workflow_steps` was a little confusing at first, maybe `step_number` or `number_of_steps` would be clearer?

Comment (nsoranzo) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: Agreed, something like `number_of_steps` would be better.

**Comment by anuprulez** [comment]
@dannon When I take latest from dev, "make client" is producing this error:

```
Warning: Uglification failed.
Unexpected token punc «)», expected punc «,». 
Line 328 in ../static/src/nls/de/locale.js
```
I committed the PR review comments but the build failed. So, I reverted the commits to the successful build and again took the latest changes from dev (that include localization changes) but it again shows the same error. Can you help? The locale.js files are present for different languages.

Thanks a lot!

**Comment by dannon** [comment]
@anuprulez Yep, taking a look and I'll get a fix out in a sec if it's busted.

(yep, found two problems, fixing them now.  once the client build is a part of the PR merge process instead of manual action, this will be easier to catch)

**Comment by dannon** [comment]
@anuprulez Fixed with #4050

**Comment by anuprulez** [comment]
@dannon thanks. It seems like a job has failed in the Travis build.

The [build](https://travis-ci.org/galaxyproject/galaxy/jobs/230760116) for [#4050](https://github.com/galaxyproject/galaxy/pull/4050) has failed for this error:
```
>> Masthead test - tabs
>> Message: beforeEach failed on tabs: Can't find variable: l
```
Travis build shows - Job #9001.4 failed

**Comment by dannon** [comment]
Yep, I was looking into that, should have a fix out in a sec.

**Comment by hexylena** [comment]
@anuprulez 

Overall:

- Love being able to quickly filter without needing to ctrl-F
- the buttons are clear
- More information in the workflow menu is really nice, it was lacking before.
- Seeing the "workflows shared with you" and "your workflows" mixed together is really nice. It goes part way toward fixing the bugs / features that still bother me in https://github.com/galaxyproject/galaxy/issues/298) 

I did find some minor bugs. Please consider testing with galaxy under a proxy_prefix (e.g. /galaxy/) as it helps detect some of these bugs.

- The three main buttons all do not respect proxy prefix, and lead to 404s (configure your workflow menu / create / upload, all go to `host/workflow/...` rather than `host/galaxy/workflow/...`)
- All options in the dropdown need to use the prefix as well

**Comment by anuprulez** [comment]
@erasche @mvdbeek @nsoranzo @dannon Thanks a lot for your reviews and help!

**Comment by guerler** [comment]
How about displaying this also when users click on 'All workflows' in the Tool panel?

**Comment by anuprulez** [comment]
@erasche Right now, I am not really sure about the Galaxy's proxy prefix how it works. I will try to explore and work on it. Thanks :)

**Comment by anuprulez** [comment]
@guerler we can do this. It will help us remove one more mako (workflow/list_for_run.mako) provided that the list of workflows displayed by /workflow and /workflow/list_for_run (All Workflows in ToolPanel) are same.

**Comment by anuprulez** [comment]
@erasche For proxy_prefix bug, I have updated all the links to contain 'Galaxy.root' before them. But I cannot test them. Does it need proxy server? Locally, they are working fine with updated url.

@guerler I updated the 'All Workflows' link in the ToolPanel to show the new workflows list. I did not remove the "list_for_run.mako" and the python code for it. I looked for "list_for_run" in the entire code repository of Galaxy, I found only two occurrences:

- client/galaxy/scripts/apps/tool-panel.js (already updated)
- lib/galaxy/webapps/galaxy/controllers/workflow.py

So, shall we remove the "list_for_run" python method in the workflow controller and the "list_for_run.mako"?

Thanks!

**Comment by hexylena** [comment]
> For proxy_prefix bug, I have updated all the links to contain 'Galaxy.root' before them. But I cannot test them. Does it need proxy server? Locally, they are working fine with updated url.

For future reference, @anuprulez it does not need a proxy server, it isn't truly for proxying, just for "run under this URL sub-path". The following galaxy.ini config changes are sufficient to test, and then you can access your galaxy at /galaxy

```
[filter:proxy-prefix]                                                                                                                              
use = egg:PasteDeploy#prefix                                                                                                                       
prefix = /galaxy                                                                                                                                   
[app:main]
filter-with = proxy-prefix                                                              
cookie_path = /galaxy                                                                                                                              
```

That said, I have tested it, it looks good now! Small visual artefact. This looks correct:

utvalg_098

But this seems to have no spacing around it:

utvalg_099

**Comment by guerler** [comment]
@anuprulez Yeah, I think we should remove it.

**Comment by anuprulez** [comment]
@erasche thanks a lot :)

Using those configuration, i could have "prefix" in the url and tested for all the actions like edit, run etc. There were couple of issues with the url again while redirecting the page from server. I have fixed those. Now all actions work fine (with and without prefix in the url).

The visuals you shared come from another mako file (import or upload workflow) and form (create workflow) from server which we have not worked upon in this PR. In future, we plan to transform those mako files as well.

Thanks a lot for your review and help!

**Comment by anuprulez** [comment]
@guerler I removed "list_for_run.mako" file and its corresponding method in the workflow controller.

**Comment by bgruening** [comment]
It's green! thanks all for your very nice comments!
Anyone wants to merge this! More to come soon :)

**Comment by jmchilton** [comment]
Really fantastic contribution and nice work! I've opened a PR with Selenium test fixes for these changes (https://github.com/bgruening/galaxy/pull/34).

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+                    _.each( workflows, function( wf ) {
+                        self.confirm_delete( self, wf );
+                    });
+                    // Register search workflow event
+                    self.search_workflow( self, self.$el.find( '.search-wf' ), self.$el.find( '.workflow-search tr' ), min_query_length );
+                }
+                else {
+                    $el_workflow.append( self._templateNoWorkflow() );
+                }
+            });
```
Comment (nsoranzo) [inline client/galaxy/scripts/mvc/workflow/workflow.js:37]: Indentation of lines 18-40 seems wrong.

Comment (dannon) [reply client/galaxy/scripts/mvc/workflow/workflow.js:37]: No, this is correct, unless I misread your concern.   The following is a single statement, not the opening of a block:
```
var self = this,
    min_query_length = 3;
```

Personally I think it's less clear than explicit var declarations, but multiple assignment like this is fine.

Comment (anuprulez) [reply client/galaxy/scripts/mvc/workflow/workflow.js:37]: The method `render` starts at column 9 
`$.getJSON ` starts at column 13 (line 18) and ends at column 13 (line 40).
Inside `$.getJSON `, lines start at column 17 and inside `if-else` lines start at column 21

File: client/galaxy/scripts/apps/analysis.js [hunk 10]
```diff
@@ -11,6 +11,7 @@ var jQuery = require( 'jquery' ),
     UserPreferences = require( 'mvc/user/user-preferences' );
     CustomBuilds = require( 'mvc/user/user-custom-builds' );
     Tours = require( 'mvc/tours' );
+    Workflows = require( 'mvc/workflow/workflow' );
```
Comment (dannon) [inline client/galaxy/scripts/apps/analysis.js]: This is made global due to the trailing semicolons above.  Fixed in https://github.com/bgruening/galaxy/pull/35/.

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 11]
```diff
@@ -0,0 +1,201 @@
+/** Workflow view */
+define( [ 'mvc/form/form-view',
+          'mvc/ui/ui-misc',
+          'mvc/ui/ui-select',
+          'mvc/tool/tool-form-composite' ], function( Form, Ui, Select, ToolForm ) {
```
Comment (dannon) [inline client/galaxy/scripts/mvc/workflow/workflow.js]: It doesn't look like we're actually using any of these here, or am I missing it?

Comment (anuprulez) [reply client/galaxy/scripts/mvc/workflow/workflow.js]: I am looking into it. Thanks!

Comment (dannon) [reply client/galaxy/scripts/mvc/workflow/workflow.js]: Sure!  Pending that, I'm happy to merge this.  Don't worry about rebuilding the client again if there are no additional changes, I can resolve that in the merge.

File: client/galaxy/scripts/mvc/workflow/workflow.js (hunk 9 above)
Comment (nsoranzo) [reply client/galaxy/scripts/mvc/workflow/workflow.js:37]: Oh how I hate this language! Sorry for the noise!

File: client/galaxy/scripts/mvc/workflow/workflow.js (hunk 11 above)
Comment (anuprulez) [reply client/galaxy/scripts/mvc/workflow/workflow.js]: Actually I rebuilt the client after removing the imports from the header of workflow.js to verify they are not needed and now I have all those static files modified. Shall I commit all these static files along with conflicting files?

### This branch has conflicts that must be resolved
```
Conflicting files
static/scripts/bundled/libs.bundled.js
static/scripts/bundled/libs.bundled.js.map
```

Comment (dannon) [reply client/galaxy/scripts/mvc/workflow/workflow.js]: That'd be fine, sure.

Comment (anuprulez) [reply client/galaxy/scripts/mvc/workflow/workflow.js]: I pushed the changes. The build is running :)

Comment (dannon) [reply client/galaxy/scripts/mvc/workflow/workflow.js]: Ahh, k, I see.  That last merge with dev wasn't necessary, but I can wait on the build and re-merge after if you want.

Comment (anuprulez) [reply client/galaxy/scripts/mvc/workflow/workflow.js]: Ok, sure. Thank you!

**Comment by dannon** [comment]
@galaxybot test this

**Comment by dannon** [comment]
(probably should have asked Travis to test this before, sorry!)

**Comment by dannon** [comment]
Thanks @anuprulez, nice work!

**Comment by anuprulez** [comment]
Thanks a lot everyone for comments and help!

**Pull request merged**